src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,438 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Reflection.Metadata;
+using System.Runtime.InteropServices;
+using Internal.IL.Stubs;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.IL
+{
+    public sealed class UnsafeAccessors
+    {
+        private const string InvalidUnsafeAccessorUsage = "Invalid usage of UnsafeAccessorAttribute.";
+        private const string AmbiguityInUnsafeAccessorUsage = "Ambiguity in binding of UnsafeAccessorAttribute.";
+
+        public static MethodIL TryGetIL(EcmaMethod method)
+        {
+            Debug.Assert(method != null);
+
+            CustomAttributeValue<TypeDesc>? decodedAttribute = method.GetDecodedCustomAttribute("System.Runtime.CompilerServices", "UnsafeAccessorAttribute");
+            if (!decodedAttribute.HasValue)
+            {
+                return null;
+            }
+
+            if (!TryParseUnsafeAccessorAttribute(method, decodedAttribute.Value, out UnsafeAccessorKind kind, out string name))
+            {
+                ThrowHelper.ThrowBadImageFormatException(InvalidUnsafeAccessorUsage);
```

Comment:
I don't remember if throwing from here will kill the compiler or just report a problem with a single method.
In general we try to NOT fail the compilation on invalid inputs if it's avoidable. For example we sometimes generate throwing method body if we can't handle the original method body - so compilation works, but at runtime it throws.
This is to handle cases where the code is not reachable at runtime, and so it would never fail on CoreCLR, but we would fail to compile it - bad.

@MichalStrehovsky would know if it's correct to throw here.

Reply:
Yeah, we probably don't guard `GetMethodIL` everywhere. This is expected to succeed. 

The best fix would be just to do something along the lines of `codestream.EmitCallThrowHelper(emit, method.Context.GetHelperEntryPoint("ThrowHelpers", "ThrowBadImageFormatException"));`.

I wouldn't bother with the exception message. The way the exception message is currently done is wrong anyway (looks like someone added an overload of `ThrowHelper.ThrowBadImageFormatException(string)` when I wasn't looking but this will not work because we need to use a localized string from CoreLib at runtime - only overloads that take an `ExceptionStringID` are good, we need to delete those that take a string).

Reply:
> The best fix would be just to do something along the lines of codestream.EmitCallThrowHelper(emit, method.Context.GetHelperEntryPoint("ThrowHelpers", "ThrowBadImageFormatException"));.

I can do that _or_ I can insert a throw to the new `ThrowAmbiguousMatchException()` that is added in this PR. I would personally prefer to match the exceptions since that is what the test is expecting.

Reply:
@MichalStrehovsky and @vitek-karas I am not emitting stubs that throw the appropriate exception at run-time. These should ensure the current tests pass as expected. I also wrapped the contents of `NativeAotILProvider.GetMethodIL()` in a `try`/`catch` and if any exceptions are thrown, I will simply return IL that calls `ThrowBadImageFormatException()`. Please let me know if this is okay.

Reply:
I have validated the tests I've written for `UnsafeAccessorAttribute` run and pass on NativeAOT.